docs_src/design/adr/system-management/0006-Metrics-Collection.md

```diff
@@ -0,0 +1,379 @@
+- an optional collection (array) of tags.  The tags are sets of key/value pairs of strings that provide amplifying information about the telemetry.  Tags may include:
+    - originating service name
+    - unit of measure associated with the telemetry value
+    - value type of the value
+    - additional values when the metric is more than just one value (example: when using a histogram, it would include min, max, mean and sum values)
+
+The metric name must be unique for that service.  Because some metrics are reported from multiple services (such as service uptime), the name is not required to be unique across all services.
+
+All information (keys, values, tags, etc.) is in string format and placed in a JSON array within the message body.  Here are some example representations:
+
+**Example metric message body with a single value**
+{"name":"service-up", "value":"120", "timestamp":"1602168089665570000", "tags":{"service":"coredata","uom":"days","type":"int64"}}
+
+**Example metric message body with multiple values**
+{"name":"api-requests", "value":"24", "timestamp":"1602168089665570001", "tags":{"service":"coredata","uom":"count","type":"int64", "mean":"0.0665", "rate1":"0.111", "rate5":"0.150","rate15":"0.111"}}
+
+!!! Note
+    The key or metric name must be unique when using go-metrics as it requires the metric name to be unique per the registry.  Metrics are considered immutable.
+
+#### REST endpoints
+
+!!! Note
+    Again, these REST endpoints are per service and meant to control which metrics for a service are turned on or off.  The REST endpoints do not provide metrics data.
+
+- Proposed endpoint for the REST endpoint that responds with what metrics the service offers is:  /api/v2/metrics
+- Body of the GET response would contain a JSON list of metrics (by metric name key) that are `on`
+- Body of the PUT request would contain a JSON list of the metrics that are to be turned `on` (others are assumed to be turned `off`)
+- Body of the PATCH request would contain a JSON list of the metrics that are to be turned `on` or `off` - leaving all other metrics unchanged
+
+#### Configuration
+- Configuration, not unlike that provided in core data or any device service, configuration will specify the message bus type and locations where the metrics messages should be sent.
+- In fact, the message bus configuration will use (or reuse if the service is already using the message bus) the common message bus configuration as defined below.
+- Metrics will be published to an /edgex/metrics/[service-name] topic where the service name will be added per service
+- Common configuration for each service for message queue configuration - inclusive of metrics:
+
+``` yaml
+[MessageQueue]
+Protocol = 'redis'  ## or 'tcp'
+Host = 'localhost'
+Port = 5573
+Type = 'redis'  ## or 'mqtt'
+PublishTopicPrefix  = 'edgex/metrics' # /<service-name> will be added to this Publish Topic prefix

```

Comment:
Device and Core data already have PublishTopicPrefix, so need an additional property for publishing metrics.
We could also add the metric name, allowing subscribers to filter for specific metrics if they wanted. This assume we publish the individual metrics separately, rather than all the service's metrics in one message.

```suggestion
MetricsPublishTopicPrefix  = 'edgex/metrics' # /<service-name>/<metric-name> will be added to this Publish Topic prefix

```

Reply:
Agreed.  Changed to /edgex/telemetry to keep it consistent with above and also added metric name.

Reply:
Also added TelemetryPublishTopicPrefix (vs PublishTopicPrefix which remains for device and core data).